finat/tensor_product.py

```diff
@@ -162,6 +162,28 @@ def point_evaluation(self, order, point, entity=None):
 
         return self._merge_evaluations(factor_results)
 
+    @property
+    def dual_basis(self):
+        # Outer product the dual bases of the factors
+        qs, pss = zip(*(f.dual_basis for f in self.factors))
+        ps = TensorPointSet(pss)
+        if len(qs) == 2:
+            # suppose there are two factors
+            qA, qB = qs
+            Aindices = gem.indices(len(qA.shape))
+            Bindices = gem.indices(len(qB.shape))
+            Q = gem.ComponentTensor(qA[Aindices]*qB[Bindices], Aindices + Bindices)
+            return Q, ps
+        else:
+            raise ValueError(f"{self.__class__.__name__} does not have a dual_basis yet!")
+
+    def dual_evaluation(self, fn):
+        if hasattr(fn, 'factors'):
```

Comment:
```
if not hasattr(fn, 'factors'):
    return super().dual_evaluation(fn)
```

Reply:
Reading the description of the interface, I am confused in that this dual_evaluation somehow does `fn(Q)` so can `fn` not return a sum-factorised gem expression?

Reply:
I'm afraid I don't quite understand your question. One never evaluates `fn(Q)`, a given function `fn` ought to eat an `AbstractPointSet`

Reply:
What I guess I mean is, surely the function should just know if it has factors and do the appropriate thing or not rather than dispatching here?

Reply:
But perhaps I just don't understand the way this fits together at the moment.

Reply:
IIRC the future implementation that @dham and I dreamed up would look something like
```python
if not hasattr(fn, 'factors'): 
    return super().dual_evaluation(fn)
else:
    assert len(fn.factors) == len(self.factors)
    for i, factor in enumerate(self.factors):
        factor_Q, factor_ps = factor.dual_basis
        factor_gem_tensor = factor.dual_evaluation(fn.factors[i])
        ...
        # somehow build up whole dual evaluation using above information
        ...
    return gem_tensor
```
 (caveat: I am not operating at 100% at the moment so I may have recalled wrong)

Reply:
OK, that looks fine. I suppose something has to dispatch and this is a reasonable place.